File: Minimal/tests/Brontide.Minimal.Interchange.Tests/Portable/PortableEncodingTests.fs
namespace Brontide.Minimal.Interchange.Tests.Portable

open NUnit.Framework
open Brontide.Minimal.Binding.Portable

/// The deterministic representation, measured against the neutral golden encodings.
///
/// The golden bytes are read from the neutral artifacts and reproduced, never restated here: a
/// codec that agreed with a hand-copied constant but not with the contract would pass a test that
/// proves nothing.
[<TestFixture>]
type PortableEncodingTests() =

    let catalog = catalogOf CoolingFixture.contract
    let limits = PortableLimits.declared

    [<Test>]
    member _.``every golden value encodes to exactly the neutral bytes``() =
        use document = readNeutral [ "vectors"; "golden-encodings.json" ]

        let entries = document.RootElement.GetProperty "encodings" |> _.EnumerateArray() |> List.ofSeq

        Assert.That(List.length entries, Is.EqualTo 6)

        for entry in entries do
            let id = entry.GetProperty("id").GetString() |> str
            let shape = readDescribedShape (entry.GetProperty "shape")
            let value = readDescribedValue (entry.GetProperty "value")
            let expected = entry.GetProperty("cbor").GetString() |> str

            let encoded =
                PortableValueCodec.encode catalog shape value
                |> Result.bind PortableCbor.encode
                |> expectOk

            Assert.That(hex encoded, Is.EqualTo expected, $"{id} did not encode to its golden bytes.")

    [<Test>]
    member _.``every golden encoding decodes back to the described value``() =
        use document = readNeutral [ "vectors"; "golden-encodings.json" ]

        for entry in document.RootElement.GetProperty("encodings").EnumerateArray() do
            let id = entry.GetProperty("id").GetString() |> str
            let shape = readDescribedShape (entry.GetProperty "shape")
            let value = readDescribedValue (entry.GetProperty "value")
            let bytes = fromHex (entry.GetProperty("cbor").GetString() |> str)

            // The Fragment the golden values attach is the one the negotiated Operation declares,
            // so decoding uses that declaration rather than an empty set.
            let decoded =
                PortableCbor.decode bytes limits
                |> Result.bind (PortableValueCodec.decode catalog shape [ CoolingFixture.hostContext ])
                |> expectOk

            Assert.That(decoded, Is.EqualTo value, $"{id} did not round-trip through the schema-guided codec.")

    [<Test>]
    member _.``deterministic key order is not ordinal string order``() =
        use document = readNeutral [ "vectors"; "golden-encodings.json" ]

        let g1 =
            document.RootElement.GetProperty("encodings").EnumerateArray()
            |> Seq.find (fun entry -> (entry.GetProperty("id").GetString() |> str) = "G1-COOLING-COMMAND-MINIMAL")

        let encoded = fromHex (g1.GetProperty("cbor").GetString() |> str)
        let text = hex encoded

        // 'loop' precedes 'enabled' because the encoded key sorts first, while ordinal string order
        // would reverse them. This is the divergence the migration obligation names.
        let loopAt = text.IndexOf "646c6f6f70"
        let enabledAt = text.IndexOf "67656e61626c6564"
        Assert.That(loopAt, Is.GreaterThanOrEqualTo 0)
        Assert.That(enabledAt, Is.GreaterThan loopAt)

    [<Test>]
    member _.``the framed example carries the declared big-endian length prefix``() =
        use document = readNeutral [ "vectors"; "golden-encodings.json" ]
        let framedExample = document.RootElement.GetProperty "framedExample"
        let framed = fromHex (framedExample.GetProperty("framed").GetString() |> str)
        let bodyBytes = framedExample.GetProperty("bodyBytes").GetInt32()

        let of' = framedExample.GetProperty("of").GetString() |> str

        let body =
            document.RootElement.GetProperty("encodings").EnumerateArray()
            |> Seq.find (fun entry -> (entry.GetProperty("id").GetString() |> str) = of')
            |> fun entry -> fromHex (entry.GetProperty("cbor").GetString() |> str)

        assertAll (fun () ->
            Assert.That(framed.Length, Is.EqualTo(PortableFraming.PrefixBytes + bodyBytes))
            shouldEqual [| 0uy; 0uy; 0uy; byte bodyBytes |] (Array.sub framed 0 4)
            shouldEqual body (Array.sub framed 4 bodyBytes))

    [<Test>]
    member _.``every rejected encoding is refused with its declared category``() =
        use document = readNeutral [ "vectors"; "golden-encodings.json" ]

        let rejected =
            document.RootElement.GetProperty "rejectedEncodings" |> _.EnumerateArray() |> List.ofSeq

        Assert.That(List.length rejected, Is.EqualTo 7)

        for entry in rejected do
            let id = entry.GetProperty("id").GetString() |> str
            let bytes = fromHex (entry.GetProperty("cbor").GetString() |> str)
            let expected = entry.GetProperty("expected").GetProperty("category").GetString() |> str

            let fault = expectRefusal (PortableCbor.decode bytes limits)

            Assert.That(
                ProtocolCategory.token fault.Category,
                Is.EqualTo expected,
                $"{id} was not refused with the category the neutral contract states."
            )

    [<Test>]
    member _.``a decimal has exactly one encoding``() =
        // The mantissa is normalized so the same fraction cannot travel two ways.
        shouldEqual (0, 123L) (expectOk (CborDecimal.normalize -1 1230L))
        shouldEqual (0, 0L) (expectOk (CborDecimal.normalize 5 0L))

    [<Test>]
    member _.``an empty frame body and an oversized prefix are separate refusals``() =
        let oversized = Array.zeroCreate<byte> (limits.MaxFrameBytes + 1)

        let fault =
            expectRefusal (PortableFraming.writeFrame (new System.IO.MemoryStream()) oversized limits System.Threading.CancellationToken.None |> _.Result)

        Assert.That(fault.Category, Is.EqualTo ProtocolCategory.LimitExceeded)

        let empty =
            expectRefusal (
                PortableFraming.writeFrame (new System.IO.MemoryStream()) Array.empty limits System.Threading.CancellationToken.None
                |> _.Result
            )

        Assert.That(empty.Category, Is.EqualTo ProtocolCategory.MalformedMessage)

File: Reference/tests/Brontide.Reference.Interchange.Tests/Portable/PortableEncodingTests.cs
using Brontide.Reference.Experimental.Binding.Portable;

namespace Brontide.Reference.Interchange.Tests.Portable;

/// <summary>
/// The Reference codec against the neutral golden encodings.
/// </summary>
/// <remarks>
/// The bytes are read from <c>binding/portable/vectors/golden-encodings.json</c> rather than
/// restated here. That is what makes them evidence: an encoder that drifts from the deterministic
/// rules fails against the same artifact the other stack will be measured against.
/// </remarks>
public sealed class PortableEncodingTests
{
    private static readonly PortableShapeCatalog Catalog =
        PortableShapeCatalog.FromContract(CoolingPortableFixture.Contract);

    [Test]
    public void Every_golden_encoding_is_reproduced_byte_for_byte()
    {
        using var document = PortableTestHarness.ReadNeutral("vectors", "golden-encodings.json");
        var encodings = document.RootElement.GetProperty("encodings").EnumerateArray().ToArray();
        Assert.That(encodings, Is.Not.Empty);

        foreach (var entry in encodings)
        {
            var id = entry.GetProperty("id").GetString()!;
            var shape = PortableTestHarness.ReadShapeReference(entry.GetProperty("shape"));
            var value = PortableTestHarness.ReadDescribedValue(entry.GetProperty("value"));
            var encoded = PortableCbor.Encode(PortableValueCodec.Encode(Catalog, shape, value));
            Assert.That(
                PortableTestHarness.Hex(encoded),
                Is.EqualTo(entry.GetProperty("cbor").GetString()),
                $"Golden encoding '{id}' was not reproduced.");
        }
    }

    [Test]
    public void Every_golden_encoding_decodes_back_to_its_described_value()
    {
        using var document = PortableTestHarness.ReadNeutral("vectors", "golden-encodings.json");
        foreach (var entry in document.RootElement.GetProperty("encodings").EnumerateArray())
        {
            var shape = PortableTestHarness.ReadShapeReference(entry.GetProperty("shape"));
            var expected = PortableTestHarness.ReadDescribedValue(entry.GetProperty("value"));
            var bytes = PortableTestHarness.FromHex(entry.GetProperty("cbor").GetString()!);
            var decoded = PortableValueCodec.Decode(
                Catalog,
                shape,
                PortableCbor.Decode(bytes, PortableLimits.Declared),
                [CoolingPortableFixture.HostContext]);
            Assert.That(
                PortableTestHarness.Hex(PortableCbor.Encode(PortableValueCodec.Encode(Catalog, shape, decoded))),
                Is.EqualTo(PortableTestHarness.Hex(PortableCbor.Encode(PortableValueCodec.Encode(Catalog, shape, expected)))),
                $"Golden encoding '{entry.GetProperty("id").GetString()}' did not round trip.");
        }
    }

    [Test]
    public void Deterministic_key_order_is_not_ordinal_string_order()
    {
        // G1 is the smallest case where the two disagree: 'loop' precedes 'enabled' because the
        // encoded key sorts first, while ordinal string order would reverse them. A codec that
        // reused the retained JSON comparer would produce the other order here.
        var encoded = PortableTestHarness.Hex(PortableCbor.Encode(PortableValueCodec.Encode(
            Catalog,
            CoolingPortableFixture.CommandV1,
            CoolingPortableFixture.Command("primary", enabled: true, requesterLabel: null))));
        Assert.That(encoded.IndexOf("6c6f6f70", StringComparison.Ordinal), Is.LessThan(
            encoded.IndexOf("656e61626c6564", StringComparison.Ordinal)));
        Assert.That(
            string.CompareOrdinal("loop", "enabled"),
            Is.GreaterThan(0),
            "Ordinal string order puts 'enabled' first, which is exactly the divergence being pinned.");
    }

    [Test]
    public void The_framed_example_carries_a_four_byte_big_endian_prefix()
    {
        using var document = PortableTestHarness.ReadNeutral("vectors", "golden-encodings.json");
        var framed = document.RootElement.GetProperty("framedExample");
        var source = document.RootElement.GetProperty("encodings").EnumerateArray()
            .Single(entry => entry.GetProperty("id").GetString() == framed.GetProperty("of").GetString());
        var body = PortableTestHarness.FromHex(source.GetProperty("cbor").GetString()!);

        Assert.That(body.Length, Is.EqualTo(framed.GetProperty("bodyBytes").GetInt32()));
        var prefix = new byte[4];
        System.Buffers.Binary.BinaryPrimitives.WriteUInt32BigEndian(prefix, (uint)body.Length);
        Assert.That(
            PortableTestHarness.Hex(prefix.Concat(body).ToArray()),
            Is.EqualTo(framed.GetProperty("framed").GetString()));
    }

    [Test]
    public void Every_rejected_encoding_fails_with_its_declared_category()
    {
        using var document = PortableTestHarness.ReadNeutral("vectors", "golden-encodings.json");
        foreach (var entry in document.RootElement.GetProperty("rejectedEncodings").EnumerateArray())
        {
            var id = entry.GetProperty("id").GetString()!;
            var bytes = PortableTestHarness.FromHex(entry.GetProperty("cbor").GetString()!);
            var expected = entry.GetProperty("expected").GetProperty("category").GetString();
            var fault = Assert.Throws<PortableFaultException>(
                () => PortableCbor.Decode(bytes, PortableLimits.Declared),
                $"Rejected encoding '{id}' decoded instead of failing.");
            Assert.That(fault!.Category.Token(), Is.EqualTo(expected), $"Rejected encoding '{id}'.");
        }
    }

    [Test]
    public void Integer_widths_use_the_shortest_form_that_represents_the_value()
    {
        Assert.Multiple(() =>
        {
            Assert.That(PortableTestHarness.Hex(PortableCbor.Encode(new CborInteger(23))), Is.EqualTo("17"));
            Assert.That(PortableTestHarness.Hex(PortableCbor.Encode(new CborInteger(24))), Is.EqualTo("1818"));
            Assert.That(PortableTestHarness.Hex(PortableCbor.Encode(new CborInteger(256))), Is.EqualTo("190100"));
            Assert.That(PortableTestHarness.Hex(PortableCbor.Encode(new CborInteger(65536))), Is.EqualTo("1a00010000"));
            Assert.That(
                PortableTestHarness.Hex(PortableCbor.Encode(new CborInteger(4294967296))),
                Is.EqualTo("1b0000000100000000"));
            Assert.That(PortableTestHarness.Hex(PortableCbor.Encode(new CborInteger(-1))), Is.EqualTo("20"));
            Assert.That(PortableTestHarness.Hex(PortableCbor.Encode(new CborInteger(-25))), Is.EqualTo("3818"));
        });
    }

    [Test]
    public void A_decimal_normalizes_so_that_one_value_has_exactly_one_encoding()
    {
        var normalized = CborDecimal.Normalize(-3, 12340);
        Assert.Multiple(() =>
        {
            Assert.That(normalized.Exponent, Is.EqualTo(-2));
            Assert.That(normalized.Mantissa, Is.EqualTo(1234));
            Assert.That(CborDecimal.Normalize(-5, 0), Is.EqualTo(new CborDecimal(0, 0)));
        });
    }
}
